### PR TITLE
opal/mca/threads: fix Argobots support

### PR DESCRIPTION
**Pull request by shintaro-iwasaki**
This PR improves the MCA/OPAL/threads/argobots support.

**Except for the first commit, this PR does not even touch anything related to the default thread (=Pthreads)**.

The first commit fixes the include header in `opal/mca/pmix/pmix_internal.h`.  Pthreads' `opal/mca/threads/threads.h` happens to indirectly include `opal_list.h` while other threading versions (e.g., Argobots) do not.  Since `opal/mca/pmix/pmix_internal.h` uses `opal_list_item_t`, `pmix_internal.h` should directly include `opal_list.h`.

The second commit introduces Argobots 1.1 static mutex/cond initializers, which has been discussed in #6578.
The third commit implements a missing function.
The fourth commit removes unused types and variables that are left somehow.
[edit] The fifth commit adds configure-time Argobots 1.1 detection.

---

This patch is related #8479, but more comprehensive.  If this PR is merged first, #8479 might need to be rebased.  If #8479 is merged first, I will rebase this PR again.

This patch passed `git clang-format`. I have tested the Argobots version on my local machine.

### CODE REVIEW

**Comment by jjhursey** [comment]
ok to test

File: opal/mca/threads/argobots/threads_argobots_wait_sync.c [hunk 1]
```diff
@@ -21,6 +21,32 @@
...
 
+void wait_sync_global_wakeup_st(int status)
+{
+    ompi_wait_sync_t *sync;
+    for (sync = wait_sync_list; sync != NULL; sync = sync->next) {
+        wait_sync_update(sync, 0, status);
+    }
+}
+
+void wait_sync_global_wakeup_mt(int status)
```
Comment (devreal) [inline opal/mca/threads/argobots/threads_argobots_wait_sync.c:32]: This code seems identical to the code in `pthreads/threads_pthreads_wait_sync.c`. Is it possible to move this code into a common subdirectory?

File: opal/mca/threads/pthreads/threads_pthreads_wait_sync.c [hunk 2]
```diff
@@ -7,6 +7,7 @@
  *                         reserved.
  * Copyright (c) 2017      IBM Corporation. All rights reserved.
  * Copyright (c) 2019      Sandia National Laboratories.  All rights reserved.
+ * Copyright (c) 2021      Argonne National Laboratory.  All rights reserved.
```
Comment (devreal) [inline opal/mca/threads/pthreads/threads_pthreads_wait_sync.c:10]: Why adding the copyright header here? I don't see any changes.

File: opal/mca/threads/pthreads/threads_pthreads_wait_sync.h [hunk 3]
```diff
@@ -15,6 +15,7 @@
  * Copyright (c) 2015-2016 Research Organization for Information Science
  *                         and Technology (RIST). All rights reserved.
  * Copyright (c) 2019      Sandia National Laboratories.  All rights reserved.
+ * Copyright (c) 2021      Argonne National Laboratory.  All rights reserved.
```
Comment (devreal) [inline opal/mca/threads/pthreads/threads_pthreads_wait_sync.h:18]: Why adding the copyright header here? I don't see any changes.

**Review by devreal — CHANGES_REQUESTED** [review]
Looks good, just a few comments/questions inline.

**Comment by devreal** [comment]
Ahh, also: since you are relying on Argobots 1.1 now would it be possible to check for the Argobots version during configure?

File: opal/mca/threads/argobots/threads_argobots_wait_sync.c (hunk 1 above)
Comment (shintaro-iwasaki) [reply opal/mca/threads/argobots/threads_argobots_wait_sync.c:33]: Thanks. Yes, we should, at least, for Argobots and Qthreads, but I also hesitate to change the Pthreads code too much. To be honest, I need someone who pushes me to change the design that can affect Pthreads. I will write my idea about this in this PR's main GitHub conversation thread.

File: opal/mca/threads/pthreads/threads_pthreads_wait_sync.c (hunk 2 above)
Comment (shintaro-iwasaki) [reply opal/mca/threads/pthreads/threads_pthreads_wait_sync.c]: You're right. I should have changed the `threads_argobots_...*.c`. I fixed it.

File: opal/mca/threads/pthreads/threads_pthreads_wait_sync.h (hunk 3 above)
Comment (shintaro-iwasaki) [reply opal/mca/threads/pthreads/threads_pthreads_wait_sync.h]: Thank you. It was my mistake. I fixed it.

**Comment by shintaro-iwasaki** [comment]
@devreal Thank you. In addition to your comments on individual files, I also added a configure-time Argobots 1.1 check (please see the last commit). I'd appreciate it if you could review this PR again.

**Comment by shintaro-iwasaki** [comment]
Regarding @devreal's comment:
> This code seems identical to the code in `pthreads/threads_pthreads_wait_sync.c`. Is it possible to move this code into a common subdirectory?

Yes I think it's a good idea.  I would like to, and at the same time, I'd like to restructure the current mca/threads code if okay.  The current design I am following is "do not modify anything related to pthreads", but as @devreal pointed out, the current one has obviously a lot of duplications.  I'd like to refactor all including a bit weird abstractions.

### Idea of refactoring

For example, I don't understand why `ompi_sync_wait_mt()` is neither `opal_thread_sync_wait_mt()` nor `opal_sync_wait_mt()`.  In my understanding, the namespace should be:
1. Everything public (except for conventional `opal_mutex` or `opal_cond`) should have a prefix of `opal_thread_`
2. Everything public should be "declared" in `threads/*.h`, not in individual threads/{pthreads, argobots, ...}/*.h` except for macros.
   - Even macros, should I do `#ifndef-#define` in `threads/*.h` and `#undef-#define` in `threads/pthreads/*.h`?
3. Everything that is not public but must be in headers (e.g., inline function) should have a special prefix to avoid global namespace pollution.

I will also remove the duplication as much as possible.

However, it also totally makes sense to stick to the current code as much as possible rather than refactoring since the current code works.  This PR is not the first time, but I'm quite new to the Open MPI development and cannot make any design decisions.  I'm happy to work on this issue once there is a consensus; I will create another PR for this.  I don't want to open a PR that is known to be rejected for the maintenance reason.

---

It depends on how much refactoring is allowed, but if we should be conservative, I would like to keep the duplications to prevent any potential interference with the pthreads layer.

**Review by devreal — APPROVED** [review]
LGTM :+1:

**Comment by devreal** [comment]
@shintaro-iwasaki From what I can see, the wait_sync implementations in all three implementations relie on three building blocks the threading implementations already provide: mutexes, conditional variables, and yield. We could just move the implementation of wait_sync into base and make use of the basic operations there. Since the thread package is selected at compile-time there should be no overhead incurred. For now, I wouldn't meddle with the naming, that could be something for later.

**Comment by shintaro-iwasaki** [comment]
@devreal Thanks for your review and approval! Can we create another PR for duplication removal? I mean, can I ask you to merge this PR before that PR? I will create that PR this week (after this PR is merged to avoid conflicts).

I understand your idea. I agree. Yes, the current `wait_sync` implementation is almost the same and we should move them to `threads/base`. If naming is not a big problem now, I will keep it.

---

I'd like to separate the PR since this change is not very trivial. Specifically, those implementations are not exactly the same, for example, the following struct directly uses `ABT_mutex`.
https://github.com/open-mpi/ompi/blob/master/opal/mca/threads/argobots/threads_argobots_wait_sync.h#L34-L35

We might need other abstractions for those raw mutex/condition variable, but it would raise concerns about performance, portability, or maintenance...

**Comment by devreal** [comment]
Absolutely, let's get this one merged and the n work on the duplication removal :+1:

**Pull request merged**